_posts/2018-09-26-Inter-service-communication.md

```diff
@@ -0,0 +1,325 @@
+To let your Spring context know that it has to search for OpenFeign clients, we just add `@EnableFeignClients`. 
+You can add this annotation to any class annotated with `@Configuration`, `@SpringBootApplication` or `@SpringCloudApplication`
+After we've enabled OpenFeign on our classpath, we can start adding OpenFeign clients. 
+When defining these clients, we have two solutions you can choose from. 
+The OpenFeign library, which provides us with the basics but very customizable OpenFeign clients, and the Spring library, that adds a few extra libraries to it for cloud tooling.
+
+## Spring
+
+{% highlight java %}
+    @FeignClient(value = "auth", fallback = FallbackAuthClient.class, configuration = FeignConfig.class)
+    public interface AuthClient { 
+        @RequestMapping(method = RequestMethod.GET, value = "checkToken")
+        boolean isTokenValid(@RequestHeader("Authorization") String token);
+    }
+{% endhighlight %}
+
+Explanation properties:
+
+* `@FeignClient`: is the annotation for Spring to recognize OpenFeign clients, OpenFeign clients have to be interfaces as it is self declarative.
+* `value/name`: is the name of the feign client that will be used to create a Ribbon load balancer which can then be linked to the target application using service discovery or a fixed list of servers. You could also use the url attribute to point your client to the target application when you're not using Ribbon.
+* `fallback`: A class where a fallback method is being implemented for resiliency. 
+
+{% highlight java %}
+    @Component
+    public class FallbackAuthClient implements AuthClient {
+        @Override
+        public boolean isTokenValid(@RequestHeader("Authorization") String token) {
+            return false;
+        }
+    }
+{% endhighlight %}
+
+* `configuration`: is for extra configuration like logging, interceptors, etc... more on that below.
+* `@RequestMapping`: Spring Cloud adds support for Spring MVC annotations and for using the same HttpMessageConverters used by default in Spring Web.
+
+## OpenFeign
+To create an OpenFeign client we need an interface and a Feign builder that tells the interface it is an OpenFeign client.
+
+{% highlight java %}
+
+public interface AuthClient {
+    @RequestLine("GET /auth")
+    @Headers({"Authorization: {token}", "Accept: application/hal+json"})
+    boolean isValid(@Param("token") String token);
+
+{% endhighlight %}
+* `@RequestLine`: is defining which verb and which uri path you are communicating to. 
+* `@Headers`: is defining the request headers that come with the request.
+
+# The builder
+OpenFeign provides us with a builder like pattern for our clients.
+When we want to customize, we just add our own customization to the builder. 
+To see the builder at work, let's create a bean of our client and return a Feign builder.
+It's important to let the builder know which client he has to target for communication. 
```

Comment:
It's important to know which **interface** the client needs to use for communication.